File: FangZhiHu/app/src/main/java/network/RxBus.java
package network;


import java.util.HashMap;

import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.processors.FlowableProcessor;
import io.reactivex.processors.PublishProcessor;
import io.reactivex.schedulers.Schedulers;


/**
 * Created by yuantongqin on 2017/8/8.
 */

public class RxBus {

    private static volatile RxBus mInstance;
    private FlowableProcessor mSubject;
    private HashMap<String, CompositeDisposable> mSubscriptionMap;

    /**
     *  PublishSubject只会把在订阅发生的时间点之后来自原始Observable的数据发射给观察者
     *  Subject同时充当了Observer和Observable的角色，Subject是非线程安全的，要避免该问题，
     *  需要将 Subject转换为一个 SerializedSubject ，上述RxBus类中把线程非安全的PublishSubject包装成线程安全的Subject。
     */
    private RxBus() {
        mSubject = PublishProcessor.create().toSerialized();
    }

    /**
     * 单例 双重锁
     * @return
     */
    public static RxBus getInstance() {
        if (mInstance == null) {
            synchronized (RxBus.class) {
                if (mInstance == null) {
                    mInstance = new RxBus();
                }
            }
        }
        return mInstance;
    }

    /**
     * 发送一个新的事件
     * @param o
     */
    public void OnEventPost(Object o) {
        mSubject.onNext(o);
    }

    //事件订阅
    public <T> Flowable<T> OnEventSubscibe(Class<T> tClass){
        return mSubject.ofType(tClass);
    }



    public <T> Disposable doSubscribe(Class<T> type, Consumer<T> next, Consumer<Throwable> error) {
        return OnEventSubscibe(type)
                .onBackpressureBuffer()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(next, error);
    }

    public void addSubscription(Object o, Disposable subscription) {

        if (mSubscriptionMap == null) {
            mSubscriptionMap = new HashMap<>();
        }
        String key = o.getClass().getName();
        if (mSubscriptionMap.get(key) != null) {
            mSubscriptionMap.get(key).add(subscription);
        } else {
            CompositeDisposable compositeSubscription = new CompositeDisposable();
            compositeSubscription.add(subscription);
            mSubscriptionMap.put(key, compositeSubscription);
        }
    }

    public void unSubscribe(Object o) {
        if (mSubscriptionMap == null) {
            return;
        }
        String key = o.getClass().getName();
        if (!mSubscriptionMap.containsKey(key)) {
            return;
        }
        if (mSubscriptionMap.get(key) != null) {
            mSubscriptionMap.get(key).dispose();
        }
        mSubscriptionMap.remove(key);
    }
}

File: FangZhiHu/CommonLibKotlin/src/main/java/network/RxBus.kt
package network

import io.reactivex.Flowable
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer
import io.reactivex.processors.FlowableProcessor
import io.reactivex.processors.PublishProcessor
import io.reactivex.schedulers.Schedulers
import java.util.HashMap

/**
 * Created by tongqinyuan on 2018/3/7.
 * 单列模式的Rxbus
 */
class RxBus private constructor(){

    private var mSubject:FlowableProcessor<Any>? = null
    private var mSubscriptionMap: HashMap<String, CompositeDisposable>? = null

    companion object {
        //单例模式
        @Volatile private var mInstance :RxBus? = null

        val getInstance : RxBus?
        get() {
            if(mInstance == null){
                synchronized(RxBus::class.java){
                    if(mInstance == null){
                        mInstance = RxBus()
                    }

                }
            }
            return mInstance
        }
    }

    init {
        mSubject = PublishProcessor.create<Any>().toSerialized()
    }

    /**
     * 发送一个事件
     */
    fun OnEventPost(any: Any){
        mSubject!!.onNext(any)
    }

    /**
     * 订阅一本杂志
     */
    fun <T> OnEventSubscibe(tClass:Class<T>) : Flowable<T> {
        return mSubject!!.ofType(tClass)
    }

    fun <T> doSubscribe(type:Class<T>,next:Consumer<T>,error:Consumer<Throwable>) : Disposable {
        return OnEventSubscibe(type)
                .onBackpressureBuffer()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(next,error)
    }

    /**
     *
    public void addSubscription(Object o, Disposable subscription) {

    if (mSubscriptionMap == null) {
    mSubscriptionMap = new HashMap<>();
    }
    String key = o.getClass().getName();
    if (mSubscriptionMap.get(key) != null) {
    mSubscriptionMap.get(key).add(subscription);
    } else {
    CompositeDisposable compositeSubscription = new CompositeDisposable();
    compositeSubscription.add(subscription);
    mSubscriptionMap.put(key, compositeSubscription);
    }
    }

    public void unSubscribe(Object o) {
    if (mSubscriptionMap == null) {
    return;
    }
    String key = o.getClass().getName();
    if (!mSubscriptionMap.containsKey(key)) {
    return;
    }
    if (mSubscriptionMap.get(key) != null) {
    mSubscriptionMap.get(key).dispose();
    }
    mSubscriptionMap.remove(key);
    }
     */


    /**
     * 添加到订阅管理
     */
    fun addSubscription(any: Any,disposable: Disposable){
        if(mSubscriptionMap == null){
            mSubscriptionMap = HashMap()
        }
        var key = any::class.java.name
        if(mSubscriptionMap!!.get(key) != null){
            mSubscriptionMap!!.get(key)!!.add(disposable)
        }else{
            var compositeSubscription = CompositeDisposable()
            compositeSubscription.add(disposable)
            mSubscriptionMap!!.put(key,compositeSubscription)
        }


    }

    /**
     * 取消订阅
     */
    fun unSubscribe(any: Any){
        if(mSubscriptionMap == null){
            return
        }
        var key = any::class.java.name
        if(!mSubscriptionMap!!.containsKey(key)){
            return
        }
        if (mSubscriptionMap!!.get(key) != null) {
            mSubscriptionMap!!.get(key)!!.dispose()
        }
        mSubscriptionMap!!.remove(key)

    }






}